snuba/query/joins/equivalence_adder.py

```diff
@@ -0,0 +1,134 @@
+
+    alias_to_entity = {
+        alias: entity_from_node(node)
+        for alias, node in from_clause.get_alias_node_map().items()
+    }
+    entity_to_alias = {}
+    for alias, entity in alias_to_entity.items():
+        assert (
+            entity not in entity_to_alias
+        ), f"Cannot process join condition. Entity {entity} is present more than once"
+        entity_to_alias[entity] = alias
+
+    column_equivalence = get_equivalent_columns(from_clause)
+    condition = query.get_condition_from_ast()
+    if condition is None:
+        return
+
+    and_components = get_first_level_and_conditions(condition)
+    conditions_to_add = []
+    for sub_condition in and_components:
+        # We duplicate only the top level conditions that reference one
+        # and only one column that has a semantic equivalent.
+        # This excludes top level conditions that contains columns from
+        # multiple entities, and cannot be pushed down to subqueries.
+        #
+        # TODO: Address top level conditions that contain multiple
+        # columns each of which has an equivalent in the same entity.
+        sole_column = _classify_single_column_condition(sub_condition, alias_to_entity)
+        if sole_column is not None:
+            for equivalent in column_equivalence.get(sole_column, []):
+                replacer = partial(
+                    _replace_col,
+                    alias_to_entity,
+                    entity_to_alias,
+                    sole_column,
+                    equivalent,
+                )
+                conditions_to_add.append(sub_condition.transform(replacer))
+
+    query.set_ast_condition(
+        combine_and_conditions([*and_components, *conditions_to_add])
+    )
+
+
+def _classify_single_column_condition(
+    condition: Expression, alias_entity_map: Mapping[str, EntityKey]
+) -> Optional[QualifiedCol]:
+    """
+    Inspects a condition to check if it is a condition on a
+    single column on a single entity
+    """
+    qualified_col: Optional[QualifiedCol] = None
+    for e in condition:
+        if isinstance(e, Column):
+            if not e.table_name:
+                return None
+            qualified = QualifiedCol(alias_entity_map[e.table_name], e.column_name)
+            if qualified_col is not None and qualified_col != qualified:
+                return None
+            qualified_col = qualified
+    return qualified_col
+
+
+def _replace_col(
+    alias_entity_map: Mapping[str, EntityKey],
+    entity_alias_map: Mapping[EntityKey, str],
+    old_col: QualifiedCol,
+    new_col: QualifiedCol,
+    expression: Expression,
+) -> Expression:
+    if not isinstance(expression, Column):
+        return expression
+
+    if (
+        expression.column_name != old_col.column
+        or expression.table_name is None
+        or expression.table_name not in alias_entity_map
+        or alias_entity_map[expression.table_name] != old_col.entity
```

Comment:
If you passed entity_alias_map[old_col.entity] to the function, then you could directly compare `expression.table_name` to that, and you wouldn't need to pass the entire `alias_entity_map` to this function.

Reply:
should not be relevant anymore